File: app/src/main/java/com/example/proif_a01_java/View/ProductListAdapter.java
package com.example.proif_a01_java.View;

import android.app.Activity;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import androidx.fragment.app.FragmentManager;

import com.example.proif_a01_java.Model.Product;
import com.example.proif_a01_java.databinding.ProductListItemBinding;


import java.util.ArrayList;

public class ProductListAdapter extends BaseAdapter {
    private Activity activity;
    private FragmentManager fm;
    private ProductListItemBinding binding;
    private ArrayList<Product> products;

    public ProductListAdapter(Activity activity, FragmentManager fm) {
        this.activity = activity;
        this.fm = fm;
        this.products = new ArrayList<>();
    }

    public void loadData (ArrayList<Product> products) {
        this.products = products;
        this.notifyDataSetChanged();
    }

    public void addProducts (ArrayList <Product> newProducts) {
        int size = newProducts.size();
        this.products.addAll(newProducts);
        this.notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return this.products.size();
    }

    @Override
    public Product getItem(int i) {
        return this.products.get(i);
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        Product currProduct =  this.getItem(i);
        if (view == null) {
            this.binding = ProductListItemBinding.inflate(this.activity.getLayoutInflater());
            view = this.binding.getRoot();
            view.setTag(this.binding);
        }else{
            this.binding = (ProductListItemBinding) view.getTag();
        }

        // SET TEXT
        this.binding.productName.setText(currProduct.name);
        this.binding.category.setText(currProduct.category);
        this.binding.condition.setText(currProduct.condition);
        this.binding.price.setText("$ " + currProduct.price);

        return binding.getRoot();
    }
}

File: app/src/main/java/com/example/proif_a01_java/View/ProductListAdapter.kt
package com.example.proif_a01_java.View


import android.app.Activity
import android.view.View
import com.example.proif_a01_java.Presenter.ProductPresenter
import android.widget.BaseAdapter
import com.example.proif_a01_java.Model.Product
import android.view.ViewGroup

import com.example.proif_a01_java.databinding.ProductListItemBinding
import java.util.ArrayList

class ProductListAdapter(private val activity: Activity, private var presenter: ProductPresenter) : BaseAdapter() {

    lateinit var binding: ProductListItemBinding
    var products: ArrayList<Product> = ArrayList()

    fun loadData(products: ArrayList<Product>) {
        this.products = ArrayList()
        this.products.addAll(products)
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return products.size
    }

    override fun getItem(i: Int): Product {
        return products[i]
    }

    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        var view: View? = view
        val currProduct = getItem(i)
        if (view == null) {
            binding = ProductListItemBinding.inflate(activity.layoutInflater)
            view = binding.root
            view.setTag(binding)
        } else {
            binding = view.tag as ProductListItemBinding
        }

        // SET TEXT
        binding.productName.text = currProduct.name
        binding.category.text = currProduct.category
        binding.condition.text = currProduct.condition
        binding.price.text = "$ " + currProduct.price
        binding.root.setOnClickListener { presenter.moveToDetails(currProduct) }
        return binding.root
    }
}